File: 1st year/hotel project c++/projectcpp/src/Room_Type_D.cpp
#include "Room_Type_D.h"
#include <iostream>
Room_Type_D::Room_Type_D()
{
    //ctor
}

Room_Type_D::~Room_Type_D()
{
    //dtor
}
Room_Type_D:: Room_Type_D(int maximum_capacity, double price_per_person, double price, int people):Room_Type_A(maximum_capacity, price_per_person, price)//kanei gia thn megisth xwrithkothta
	{
	    //gia thn timh ana atomo kai gia thn timh ana hmera oti kai o constructor ths Room_Type_A
	minimum_people = people;////thetei ton elaxisto arithmo atomwn apo to orisma
}
bool Room_Type_D::new_booking(Booking* booketo) { //elexgei an ginetai kainouria krathsh kai thn kanei
	if (maximum_capacity < booketo->getPeople() && minimum_people>booketo->getPeople()) {
			//elegxei an threitai h megisth xwritikothta kai o elaxistos arithmos hmerwn kai atomwn
			return false;//epistrefei false an den threitai
		}
	if (maximum_capacity < booketo->getPeople()) {//elegxei an h xwrithkothta einai mikroterh apo ta atoma
		return false;
	} //an einai epistrefei false kai termatizei
	if(booketo->getArrival()+booketo->getDays_of_accomodation()-1<=30){
		for (int j = booketo->getArrival(); j <= (booketo->getArrival() + booketo->getDays_of_accomodation() - 1); j++)
			//gia j apo thn mera afiksh mexri thn hmera pou tha fygoun(hmera afikshs+ meres pou menoun)
			if (availability[j] != 0) {//an h diahesimothta ekeinh thn hmera den einai kenh (exei yparksei krathsh se auto to dwmatio gia autes ths meres dhladh
				return false;
			}//epistrefei false kai termatizei

		for (int j = booketo->getArrival(); j <= (booketo->getArrival() - 1 + booketo->getDays_of_accomodation()); j++)
			//to j einai idio me to j apo thn prohgoumenh for
			availability[j] =booketo; //bazei se kathe keli ths diathesimothtas tou dwmatiou ekeinou tou dwmatiouta stoixeia ths krathshs
		booketo->setRoom(this);//thetei thn anafora thw krathshs se auto to dwmatio
		return true;  //epistrefei true kai termatizei

	}
	else{
		cout << "Out of Range error: " <<endl;
	//uparxei periptwsh na dothoun meres pou den einai sta oria
														//tou pinaka diathesimothtas
		return false;//tote epistrefei false
	}}
//this type of room cannot cancel its booking
bool Room_Type_D:: cancelation(int cancelID)//yperkalyptei thn methodo akurwshs wste na mhn ginetai
{
	cout<<"Cancelling is not possible with this kind of room"<<endl;
	bool control = false;
	return control;
}

File: 1st year/hotel project java/project/Room_Type_D.java
 

public class Room_Type_D extends Room_Type_A {
    private int minimum_people;//elaxista atoma
    Room_Type_D(){
    
    }
    Room_Type_D(int maximum_capacity,double price_per_person,double price,int people){
       super(maximum_capacity,price_per_person,price);//kanei gia thn megisth xwrithkothta 
    // gia thn timh ana atomo kai gia thn timh ana hmera oti kai o constructor ths Room_Type_A
        this.minimum_people=people;////thetei ton elaxisto arithmo atomwn apo to orisma
       }
    public int getMinimum_people() {
        return minimum_people;
    }

    public void setMinimum_people(int minimum_people) {
        this.minimum_people = minimum_people;
    }
    //To book this type of room you have to be more than a minimum number of people 
    boolean new_booking(Booking booketo)
    {
        if(this.maximum_capacity < booketo.getPeople()&& minimum_people>booketo.getPeople()){
             //elegxei an threitai h megisth xwritikothta kai o elaxistos arithmos atomwn 
            return false;//an den threitai epistrefei false
        } 
        //alliws kanei oti kai h new_booking
        try{
         for(int j = booketo.getArrival();j <=(booketo.getArrival() + booketo.getDays_of_accomodation()-1);j++) 
         if(this.availability[j] != null){
            return false;}                                                  
         
          for(int j = booketo.getArrival();j <=(booketo.getArrival()-1 + booketo.getDays_of_accomodation());j++) 
          this.availability[j] = booketo;      
          //System.out.println("sdjfiofgohsigokhaoi");
          booketo.setRoom(this);
          return true;  
             
            }
        catch(java.lang.ArrayIndexOutOfBoundsException e){
            return false;
        }
    }
    //this type of room cannot cancel its booking
    boolean cancelation(int cancelID)//yperkalyptei thn methodo akurwshs wste na mhn ginetai
    {
        System.out.println("Cancelling is not possible with this kind of room");
        boolean control=false;
        return control;
    }
}

